dali/test/python/test_operator_normalize.py

```diff
@@ -0,0 +1,255 @@
+
+    if stddev is None:
+        stddev = batch_stddev(in_batch, axes, mean)
+
+    out = []
+    for x in in_batch:
+        with np.errstate(divide='ignore', invalid='ignore'):
+            norm = (x - mean) / stddev
+        out.append(np.nan_to_num(norm, copy = False, nan = 0, posinf = 0, neginf = 0))
+    return out
+
+# Generate random tensors with given dimensionality.
+# If batch_norm is True, the extents in non-reduced axes are equal.
+# If no using batch_norm, axes argument is ignored.
+def generate_data(dims, batch_size, batch_norm, axes):
+    shapes = np.random.randint(1, 10, [batch_size, dims], dtype=int)
+    if batch_norm and axes is not None:
+        for i in range(1, batch_size):
+            for a in range(dims):
+                if a not in axes:
+                    shapes[i, a] = shapes[0, a]
+    shapes = shapes.tolist()
+    return [np.random.rand(*s).astype(np.float32) for s in shapes]
+
+def custom_mean(batch_norm, axes):
+    bias = 0.3  # make the result purposefully slightly off
+    if type(axes) is list:
+        axes = tuple(axes)
+    if batch_norm:
+        def whole_batch_mean(batch):
+            out = batch_mean(batch, axes) + bias
+            return [[out.astype(np.float32) for _ in range(len(batch))]]
+        return whole_batch_mean
+    else:
+        def per_sample_mean(batch):
+            return [[x.mean(axis = axes, keepdims = True, dtype=np.float32) + bias for x in batch]]
+        return per_sample_mean
+
+def custom_stddev(batch_norm, axes):
+    bias = 1.3  # make the result purposefully slightly off
+    mean_func = custom_mean(batch_norm, axes)
+    if type(axes) is list:
+        axes = tuple(axes)
+    if batch_norm:
+        def whole_batch_stddev(batch):
+            mean = mean_func(batch)[0][0]
+            out = bias * batch_stddev(batch, axes, mean)
+            return [[out for _ in range(len(batch))]]
+        return whole_batch_stddev
+    else:
+        def per_sample_stddev(batch):
+            mean = mean_func(batch)[0]
+            out = []
+            for i in range(len(batch)):
+                stddev = bias * np.sqrt(((batch[i] - mean[i])**2).mean(axis = axes, keepdims = True))
+                out.append(stddev)
+            return [out]
+        return per_sample_stddev
+
+def normalize_list(whole_batch, data_batch, axes = None, mean = None, stddev = None):
+    if whole_batch:
+        return batch_norm(data_batch, axes, mean, stddev)
+    else:
+        if type(mean) is not list:
+            mean = [mean] * len(data_batch)
+        if type(stddev) is not list:
+            stddev = [stddev] * len(data_batch)
+        return [normalize(data_batch[i], axes, mean[i], stddev[i]) for i in range(len(data_batch))]
+
+def err(l1, l2):
+    return np.max([np.max(np.abs(a[0] - a[1])) for a in zip(l1, l2)])
+
+def check(l1, l2):
+    for a in zip(l1, l2):
+        np.allclose(a[0], a[1], rtol=1e-3, atol=1e-3)
+
+class NormalizePipeline(Pipeline):
```

Comment:
Is it possible to split this mega pipeline into a couple of smaller one to make test cases decoupled one from another?

Reply:
This is actually one case - i.e. it works on one input. There are already _hundreds_ of tests using this pipeline, and, at the very least, I'm pleased to say that they don't take ages.

Reply:
...but I'd like to hear some advice on _how_ to do it without too much copy/pasting.

Reply:
There is no good advice, only complain